solutions/USACO open16-bull-in-china-shop.mdx

```diff
@@ -0,0 +1,135 @@
+---
+id: usaco-640
+source: USACO Bronze 2016 Open
+title: Bull in a China Shop
+author: Maggie Liu
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_bcs_bronze_open16.html)
+
+## Explanation
+
+We can loop through all possible pairs of pieces and shift them to check if they can make the original figurine. Since we cannot shift a piece so any '#' fall outside the $N \times N$ grid, we need to find the sides of the pieces to determine how far to shift them.
+
+<Warning>
+
+The official solution doesn't check if any '#' fall outside the $N \times N$ grid, which is wrong.
+
+</Warning>
+
+As we read in the pieces, we find the left, right, top and bottom sides of the piece and store this in the $\texttt{side}$ array. When we are shifting, we shift $\texttt{piece[i]}$ $\texttt{idy}$ units to the left and $\texttt{idx}$ units up, and shift $\texttt{piece[j]}$ $\texttt{jdy}$ units to the left and $\texttt{jdx}$ units up. This will send the piece at $(x + \texttt{idx}, y + \texttt{idy})$ to $(x, y)$ and the piece at $(x + \texttt{jdx}, y + \texttt{jdy})$ to $(x, y)$. So, we see that $\texttt{idy}$ must be bounded by $\texttt{right} - n$ and $\texttt{left}$, and $\texttt{idx}$ must be bounded by $\texttt{bottom} - n$ and $\texttt{top}$, with $\texttt{jdy}$ and $\texttt{jdx}$ bounded similarly.
+
+Once we have shifted the pieces, if there are two '#' in the same place, or if the result of superimposing the pieces is not the same as the original figurine ($\texttt{pieces[0]}$), then this shift will not work.
+
```

Comment:
This too. It is already explained in the editorial. Just put something like: We will do the same as the editorial. But in order to avoid shifting '#' pieces outside de N * N grid, we will save the maximum shifts we can do in one direction with every piece.